sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/CustomShuffleReaderExec.scala

```diff
@@ -62,20 +61,113 @@ case class CustomShuffleReaderExec private(
-    if (cachedShuffleRDD == null) {
-      cachedShuffleRDD = child match {
-        case stage: ShuffleQueryStageExec =>
-          new ShuffledRowRDD(
-            stage.shuffle.shuffleDependency, stage.shuffle.readMetrics, partitionSpecs.toArray)
-        case _ =>
-          throw new IllegalStateException("operating on canonicalization plan")
+  def hasSkewedPartition: Boolean = {
+    partitionSpecs.exists(_.isInstanceOf[PartialReducerPartitionSpec])
+  }
+
+  def isLocalReader: Boolean = {
+    if (partitionSpecs.exists(_.isInstanceOf[PartialMapperPartitionSpec])) {
+      assert(partitionSpecs.forall(_.isInstanceOf[PartialMapperPartitionSpec]))
+      true
+    } else {
+      false
+    }
+  }
+
+  private def shuffleStage = child match {
+    case stage: ShuffleQueryStageExec => Some(stage)
+    case _ => None
+  }
+
+  override lazy val metrics = {
+    if (shuffleStage.isDefined) {
+      val numPartitions = SQLMetrics.createMetric(sparkContext, "number of partitions")
+      numPartitions.set(partitionSpecs.length)
+      Map("numPartitions" -> numPartitions) ++ {
+        if (isLocalReader) {
+          // We split the mapper partition evenly when creating local shuffle reader, so no
+          // data size info is available.
+          Map.empty
+        } else {
+          val maxSize = SQLMetrics.createSizeMetric(sparkContext, "max data size of partitions")
+          val minSize = SQLMetrics.createSizeMetric(sparkContext, "min data size of partitions")
+          val avgSize = SQLMetrics.createSizeMetric(sparkContext, "avg data size of partitions")
+          val mapStats = shuffleStage.get.mapStats.bytesByPartitionId
+          val sizes = partitionSpecs.map {
+            case CoalescedPartitionSpec(startReducerIndex, endReducerIndex) =>
+              startReducerIndex.until(endReducerIndex).map(mapStats(_)).sum
+            case p: PartialReducerPartitionSpec => p.dataSize
+            case p => throw new IllegalStateException("unexpected " + p)
+          }
+          maxSize.set(sizes.max)
+          minSize.set(sizes.min)
+          avgSize.set(sizes.sum / sizes.length)
+          Map(
+            "maxPartitionDataSize" -> maxSize,
+            "minPartitionDataSize" -> minSize,
+            "avgPartitionDataSize" -> avgSize)
+        }
+      } ++ {
+        if (hasSkewedPartition) {
+          val skewedPartitions = SQLMetrics.createMetric(
+            sparkContext, "number of skewed partitions")
+          var i = 0
+          var currentReducerIndex = -1
+          while (i < partitionSpecs.length) {
+            partitionSpecs(i) match {
+              case p: PartialReducerPartitionSpec =>
+                if (p.reducerIndex != currentReducerIndex) {
+                  skewedPartitions.add(1)
+                  currentReducerIndex = p.reducerIndex
+                }
+              case _ => currentReducerIndex = -1
+            }
+            i += 1
+          }
```

Comment:
Maybe you're considering perf here, but something like:

```
partitionSpecs
  .filter(_.isInstanceOf[PartialReducerPartitionSpec])
  .map(_.reducerIndex).distinct.sum
```

could be more readable?